### PR TITLE
Remove Pymatgen dependency, use ASE to read/write XYZ files

### LINKED ISSUE #33: relaxed_structure StructureData should preserve atom ordering
It seems that the output node `relaxed_structure` of optimization runs has a different atom order that the original input structure.
I need to do a bit more testing, but probably this line from the parser needs tweaking.

https://github.com/pzarabadip/aiida-orca/blob/5b2cba2b518837c35179b52ac1141eda27609f4b/aiida_orca/parsers/__init__.py#L86

### PR DESCRIPTION
**Pull request by danielhollas**
Split from #46 

The Pymatgen is quite a heavy dependency.  Due to unrelated issues in AiiDAlab environment, I had problems with Pymatgen. I looked around and we could get rid of the Pymatgen dependency (and in turn many other transitive dependencies from `aiida[atomic_tools]`) and use ASE instead for parsing. ASE seems to have much less dependencies,

```console
$ pip show ase
Name: ase
Version: 3.19.3
Requires: matplotlib, numpy, scipy
```

```console
$ pip show pymatgen
Name: pymatgen
Version: 2022.2.1
Summary: Python Materials Genomics is a robust materials analysis code that defines core object representations for structures and molecules with support for many electronic structure codes. It is currently the core analysis code powering the Materials Project (https://www.materialsproject.org).
Requires: Cython, matplotlib, monty, networkx, numpy, palettable, pandas, plotly, pybtex, requests, ruamel.yaml, scipy, spglib, sympy, tabulate, tqdm, uncertainties
```

In addition, this PR makes the example scripts in `examples/` compatible with AiiDA-2.0 (while preserving compatibility with AiiDA-1.x).

**NOTE:** Reference data for the parser test needed to be updated.
Apparently, Pymatgen was not preserving the order of atoms when reading the XYZ structure from file, and also seemed
to have been shifting the coordinates (translating and/or rotating to molecule). I have already reported this in #33. This PR fixes #33, ASE parser preserves the atom ordering and coordinates exactly as they are, besides minor rounding differences.

### CODE REVIEW

File: tests/parsers/test_orca/test_orca_default.yml [hunk 1]
```diff
@@ -471,44 +471,44 @@ output_parameters:
...
   sites:
-  - kind_name: H
+  - kind_name: C
     position:
-    - 6.902699000455
-    - 5.5216858017195
-    - 5.8053320967323
+    - 5.6454888132719
+    - 5.8099486645379
+    - 5.6434676576406
```
Comment (sphuber) [inline tests/parsers/test_orca/test_orca_default.yml:494]: Should it worry us that these positions do not match what `pymatgen` parsed? Or was the input changed somehow?

File: aiida_orca/calculations/orca_orca.py [hunk 2]
```diff
@@ -123,24 +123,11 @@ def prepare_for_submission(self, folder: Folder) -> CalcInfo:
...
-
-#EOF
+    def _write_structure(structure: StructureData, folder: Folder, filename: str) -> None:
+        """Function that writes a structure to a file in an XYZ format"""
+
+        # create file with the XYZ structure
+        ase_struct = structure.get_ase()
+        # ORCA cannot read the extended XYZ format, hence plain=True is needed.
+        # https://wiki.fysik.dtu.dk/ase/ase/io/formatoptions.html#ase.io.extxyz.write_extxyz
+        ase_struct.write(folder.get_abs_path(filename), format='extxyz', plain=True)
```
Comment (sphuber) [inline aiida_orca/calculations/orca_orca.py:133]: If extended-xyz is not supported, why not use xyz format directly?
```suggestion
        ase_struct.write(folder.get_abs_path(filename), format='xyz')
```

File: aiida_orca/parsers/__init__.py [hunk 3]
```diff
@@ -75,11 +74,17 @@ def _remove_nan(parsed_dictionary: dict) -> dict:
...
+            #with out_folder.base.repository.open(fname_relaxed) as handler:
+            with out_folder.open(fname_relaxed) as handler:
+                ase_structure = ase.io.read(handler, format='xyz', index=0)
+                if not ase_structure:
+                    self.logger.error(f'Could not read structure from output file {fname_relaxed}')
+                    return self.exit_codes.ERROR_OUTPUT_PARSING
+                # Temporary hack to support AiiDA 1.x, which needs default cell
+                # even for non-periodic structures.
+                ase_structure.set_cell([1.0, 1.0, 1.0])
+                relaxed_structure = StructureData(ase=ase_structure)
```
Comment (sphuber) [inline aiida_orca/parsers/__init__.py]:
```suggestion
            if not ase_structure:
                self.logger.error(f'Could not read structure from output file {fname_relaxed}')
                return self.exit_codes.ERROR_OUTPUT_PARSING
            # Temporary hack to support AiiDA 1.x, which needs default cell
            # even for non-periodic structures.
            ase_structure.set_cell([1.0, 1.0, 1.0])
            relaxed_structure = StructureData(ase=ase_structure)
```

File: examples/example_0.py [hunk 4]
```diff
@@ -36,22 +39,24 @@ def example_opt(orca_code, nproc, submit=True):
                     'nproc': nproc,
                 }
             },
-            'input_keywords': ['PBE', 'SV(P)', 'Opt'],
+            'input_keywords': ['PBE', 'def2-SVP', 'Opt'],
```
Comment (sphuber) [inline examples/example_0.py:42]: why is this change necessary?

File: aiida_orca/parsers/__init__.py [hunk 5]
```diff
@@ -75,11 +74,17 @@ def _remove_nan(parsed_dictionary: dict) -> dict:
...
         if parsed_dict.get('optdone', False):
-            with out_folder.open(fname_relaxed, 'rb') as handle:
-                with tempfile.NamedTemporaryFile('w+b', suffix=pathlib.Path(fname_relaxed).suffix) as tmpfile:
-                    shutil.copyfileobj(handle, tmpfile)
-                    tmpfile.flush()
-                    relaxed_structure = StructureData(pymatgen_molecule=Molecule.from_file(tmpfile.name))
+            # Change this when we drop AiiDA 1.x support
+            #with out_folder.base.repository.open(fname_relaxed) as handler:
+            with out_folder.open(fname_relaxed) as handler:
+                ase_structure = ase.io.read(handler, format='xyz', index=0)
```
Comment (sphuber) [inline aiida_orca/parsers/__init__.py]: Nitpick alert: usually it is referred to as a file-handle
```suggestion
            with out_folder.open(fname_relaxed) as handle:
                ase_structure = ase.io.read(handle, format='xyz', index=0)
```

File: examples/example_1.py [hunk 6]
```diff
@@ -25,12 +25,14 @@ def example_opt_restart(orca_code, nproc, submit=True, opt_calc_pk=None):
...
+    ase_struct = ase.io.read(xyz_path, format='xyz', index=0)
+    ase_struct.set_cell([1.0, 1.0, 1.0])
+    structure = StructureData(ase=ase_struct)
 
     # old gbw file
     retr_fldr = load_node(opt_calc_pk).outputs.retrieved
-    with retr_fldr.open('aiida.gbw') as handler:
-        gbw_file = SinglefileData(handler.name)
+    with retr_fldr.open('aiida.gbw', 'rb') as handler:
+        gbw_file = SinglefileData(handler)
```
Comment (sphuber) [inline examples/example_1.py]:
```suggestion
    with retr_fldr.open('aiida.gbw', 'rb') as handle:
        gbw_file = SinglefileData(handle)
```

File: examples/example_4.py [hunk 7]
```diff
@@ -28,38 +28,48 @@ def example_simple_tddft(orca_code, nproc, submit=True, opt_calc_pk=None):
...
 
     # Construct process builder
     builder = OrcaCalculation.get_builder()
 
-    builder.structure = load_node(opt_calc_pk).outputs.relaxed_structure
+    # old gbw file
+    opt_calc = load_node(opt_calc_pk)
+    retr_fldr = opt_calc.outputs.retrieved
+    with retr_fldr.open('aiida.gbw', 'rb') as handler:
+        gbw_file = SinglefileData(handler)
```
Comment (sphuber) [inline examples/example_4.py]: This is a side note, but seeing this, would it maybe make sense to have the parser always attach `aiida.gbw` as a `SinglefileData` output node, instead of storing it in the retrieved folder? Given that the `OrcaCalculation` takes this as an input for restarts? Would simplify user code at least and also make the provenance clearer

**Review by sphuber — CHANGES_REQUESTED** [review]
Thanks @danielhollas . Looks good to me, just some minor comments/suggestions

File: examples/example_4.py (hunk 7 above)
Comment (danielhollas) [reply examples/example_4.py]: Interesting idea, but I don't think this should be the default, since these files can get huge since they are storing the whole wavefunction. Maybe we could have a calcjob Bool parameter that would determine if to store this file or not. Not sure if there is a precedent in AiiDA workflows for this.

File: examples/example_0.py (hunk 4 above)
Comment (danielhollas) [reply examples/example_0.py:42]: Not necessary at all, I think I was just trying to harmonize the used basis set across the files (and `def2-SVP` is a more modern version so it's better to use it as an example I think). (in general there are a bunch of cleanups in these files that I did, and separating them out would be too much extra work for me)

File: aiida_orca/calculations/orca_orca.py (hunk 2 above)
Comment (danielhollas) [reply aiida_orca/calculations/orca_orca.py:133]: Good questions. The ASE docs are quite confusing, but I believe this is the first thing I tried and it didn't work. I think it's because internally, even if you specify `xyz`, it calls out the routines from `extxyz` anyway.

File: tests/parsers/test_orca/test_orca_default.yml (hunk 1 above)
Comment (danielhollas) [reply tests/parsers/test_orca/test_orca_default.yml:494]: I suspect Pymatgen was re-orienting the molecule somehow, but I will double check.

File: examples/example_0.py (hunk 4 above)
Comment (sphuber) [reply examples/example_0.py:42]: Ok. fine to keep then

File: examples/example_4.py (hunk 7 above)
Comment (sphuber) [reply examples/example_4.py]: But you are anyway storing them already as they are part of the `retrieved` node, which is also stored locally. So either this should not be retrieved by default (and a switch in the inputs can let a user tell the plugin to retrieve it when needed) or it might as well be attached as a node, which wouldn't change the load on the database.

File: aiida_orca/parsers/__init__.py (hunk 5 above)
Comment (danielhollas) [reply aiida_orca/parsers/__init__.py]: Good point, I was confused by this before. I've applied your suggestion throughout.

File: examples/example_4.py (hunk 7 above)
Comment (danielhollas) [reply examples/example_4.py]: Okay, so to double check I understand your suggestion, you would attach this file directly as a `SinglefileData`, and removed it from the retrieved folder? I am not sure about the details of AiiDA backed, is SinglefileData essentially just a link to a file storage (so the means of how these files were stored would be identical in the backend?

One thing I am worried about, I am not sure if it is possible now to delete files from the retrieved folder, without AiiDA complaining about provenance? If that is the case, I think the current design is good since users might want to delete these files to save space. If this files were attached dirrectly as output nodes, I don't think one could easily get rid of them, right?

(in any case, this is a discussion for another PR and should not block this one)

Comment (sphuber) [reply examples/example_4.py]: > I am not sure about the details of AiiDA backed, is SinglefileData essentially just a link to a file storage (so the means of how these files were stored would be identical in the backend?

Essentially it is almost the same. In one case you have a `FolderData` node and in the other a `SinglefileData` node. The former can store any number of files, but the `SinglefileData` just the one. The files itself are essentially stored in the same way. It is just that if a restart calculation requires this file to be passed as a `SinglefileData`, it would make sense to already have the parser perform this operation instead of forcing the user to always have this code snippet to do it (which also duplicates the file in a sense).

> One thing I am worried about, I am not sure if it is possible now to delete files from the retrieved folder, without AiiDA complaining about provenance? If that is the case, I think the current design is good since users might want to delete these files to save space. If this files were attached dirrectly as output nodes, I don't think one could easily get rid of them, right?

The `retrieved` output is a `FolderData` node which is, just like all other nodes, immutable so you cannot delete a file. So this would not be a problem. If it really is a problem that this file is always retrieved and stored (as is currently the case) you could change the code to not retrieve it by default, but have an input to the calculation that will trigger to retrieve it. Historically in `aiida-quantumespresso` this was done through the `settings` input node, which could be used to change the behavior of the calculation. You could think of adding something like `retrieve_gbw`. The `OrcaCalculation` could check this key in the `settings` node, and when set to `True` add the `aiida.gbw` file to the `retrieve_temporary_list`. The `OrcaParser` can then wrap the file from the retrieved temporary folder in a `SinglefileData` node.

File: tests/parsers/test_orca/test_orca_default.yml (hunk 1 above)
Comment (danielhollas) [reply tests/parsers/test_orca/test_orca_default.yml:494]: Just noting that the coordinates that are parsed now with ASE correspond exactly to the reference XYZ file.

Comment (sphuber) [reply tests/parsers/test_orca/test_orca_default.yml:494]: If it makes sense than this comment can be considered resolved.

File: examples/example_4.py (hunk 7 above)
Comment (sphuber) [reply examples/example_4.py]: Would be happy to make a PR for this to show the concept if you and @pzarabadip think it makes sense to have a switch to retrieve (and potentially attach as `SinglefileData`) the file.

Comment (danielhollas) [reply examples/example_4.py]: Thank you for the explanation! Yes, I think having it be optional definitely makes sense.

> Historically in aiida-quantumespresso this was done through the settings input node, which could be used to change the behavior of the calculation. You could think of adding something like retrieve_gbw. The OrcaCalculation could check this key in the settings node, and when set to True add the aiida.gbw file to the retrieve_temporary_list. The OrcaParser can then wrap the file from the retrieved temporary folder in a SinglefileData node.

Yep, that makes sense for sure. The only question would than be what should be the default. Just to make sure, if the file is both in SinglefileData node and FolderData, it won't be duplicated then? (I guess in AiiDA 2.0 which automaticaly deduplicates files that makes sense, not sure for AiiDA 1.x).

> Would be happy to make a PR for this to show the concept if you and @pzarabadip think it makes sense to have a switch to retrieve (and potentially attach as SinglefileData) the file.

Thanks, much appreciated! I think for now I'd like to focus on merging the existing PRs and then we can come back to this. I definitely think it's a good idea. Maybe for now you could open an issue and just copy over the info that you provided above?

Comment (sphuber) [reply examples/example_4.py]: > Just to make sure, if the file is both in SinglefileData node and FolderData, it won't be duplicated then? (I guess in AiiDA 2.0 which automaticaly deduplicates files that makes sense, not sure for AiiDA 1.x).

No, I would keep it either in the `FolderData` or just as the `SinglefileData`, not in both. In 1.x it would indeed duplicate the file in the repository. In 2.x it would indeed be automatically deduplicated, but that won't be the case for all storage backends (which are pluginnable as of v2.0).

I just read a bit on ORCA and it seems that the .gbw file is the wavefunction. It looks to me that maybe it should not be retrieved by default. We do the same with `aiida-quantumespresso`. The wavefunctions are never retrieved because they are huge. If you want to restart, you simply use the `remote_folder` of the original calculation. I think we should apply this approach for this plugin as well. We can then add the option to retrieve it as a single file in case people want to store a wavefunction locally. This can be useful if the calculation needs to be restartable even in a long time when the scratch on the remote will most likely already have been cleaned.

But I agree, this all not necessary for this PR.

Comment (ezpzbz) [reply examples/example_4.py]: I agree with making the retrieve optional upon user request. I had in mind (at the time) to use it for conversion to `wfx` file to run `QTAIM` calculation but have not done it after all. Either way, it should be easily doable with `remore_folder` too.

File: examples/example_5.py [hunk 8]
```diff
@@ -51,15 +48,18 @@ def example_opt(orca_code, nproc, submit=True):
 
     builder.orca.metadata.options.resources = {
         'num_machines': 1,
-        'num_mpiprocs_per_machine': 1,
+        'num_mpiprocs_per_machine': nproc,
     }
     builder.orca.metadata.options.max_wallclock_seconds = 1 * 10 * 60
     if submit:
-        print('Testing Orca Opt Calculations...')
+        print('Testing basic Orca workflow...')
```
Comment (ezpzbz) [inline examples/example_5.py]: @danielhollas maybe just changing this to something like `Testing OrcaBaseWorkChain` for clarity.

**Review by ezpzbz — APPROVED** [review]
It's good to go on my side. Feel free to merge when tiny remaining bits are done.

File: examples/example_5.py (hunk 8 above)
Comment (danielhollas) [reply examples/example_5.py]: Thanks, done.
